File: services/constraint-engine/sim.go
package main

import (
	"fmt"
	"math"
	"strings"
)

func analyzeFix(fix string) Capabilities {
	f := strings.ToLower(fix)
	has := func(subs ...string) bool {
		for _, s := range subs {
			if strings.Contains(f, s) {
				return true
			}
		}
		return false
	}
	return Capabilities{
		Cache:      has("redis", "cache", "memcached"),
		Persistent: has("redis", "persist", "external", "expire", "ttl", "database", "db"),
		Atomic:     has("lua", "incr", "atomic", "transaction", "watch", "compare-and"),
		Pooling:    has("pool", "pooling"),
		Replicas:   has("replica", "read replica", "sentinel"),
		Horizontal: has("lb", "load balancer", "load-balancer", "scale out", "scale-out", "horizontal", "multiple servers", "autoscal"),
		Sharded:    has("shard", "consistent hash", "consistent-hash", "partition"),
	}
}

func requirementCheck(level ConstraintLevel, fix string) (bool, string) {
	f := strings.ToLower(fix)
	containsAll := func(keys []string) (bool, string) {
		for _, k := range keys {
			if !strings.Contains(f, strings.ToLower(k)) {
				return false, k
			}
		}
		return true, ""
	}
	containsAny := func(keys []string) bool {
		for _, k := range keys {
			if strings.Contains(f, strings.ToLower(k)) {
				return true
			}
		}
		return false
	}
	if len(level.Requires) > 0 {
		if ok, missing := containsAll(level.Requires); !ok {
			return false, fmt.Sprintf("missing required change: %q (need %s)", missing, level.ChangeType)
		}
	}
	if len(level.RequiresAny) > 0 {
		if !containsAny(level.RequiresAny) {
			return false, fmt.Sprintf("fix doesn't address the constraint (need one of: %s)", strings.Join(level.RequiresAny, ", "))
		}
	}
	return true, ""
}

func simulate(level ConstraintLevel, caps Capabilities) SimulationResult {
	rps := float64(level.TargetRPS)
	const singleNodeCap = 2000.0
	effectiveCap := singleNodeCap
	if caps.Horizontal {
		effectiveCap *= 6
	}
	if caps.Sharded {
		effectiveCap *= 1.5
	}
	if caps.Replicas {
		effectiveCap *= 1.3
	}

	util := rps / effectiveCap
	cpu := math.Min(100, util*100)

	base := 40.0
	if caps.Cache {
		base = 8.0
	}
	if caps.Pooling {
		base -= 4
	}
	if base < 3 {
		base = 3
	}

	queueFactor := 1.0
	if util < 0.95 {
		queueFactor = 1.0 / (1.0 - util)
	} else {
		queueFactor = 20.0 + (util-0.95)*200.0
	}

	p50 := base * math.Min(queueFactor, 6)
	p99 := base * queueFactor * 2.5

	errRate := 0.0
	if util > 1.0 {
		errRate += (util - 1.0) * 60.0
	}
	if util > 0.85 {
		errRate += (util - 0.85) * 10.0
	}
	errRate = math.Min(errRate, 95.0)

	throughput := int(math.Min(rps, effectiveCap))

	return SimulationResult{
		P50Latency: round1(p50),
		P99Latency: round1(p99),
		ErrorRate:  round1(errRate),
		Throughput: throughput,
		CPUPct:     round1(cpu),
	}
}

func evaluate(level ConstraintLevel, fix string) SimulationResult {
	caps := analyzeFix(fix)
	res := simulate(level, caps)

	if ok, reason := requirementCheck(level, fix); !ok {
		res.Passed = false
		res.FailureReason = reason
		return res
	}

	switch {
	case res.P99Latency > float64(level.LatencySLA):
		res.Passed = false
		res.FailureReason = fmt.Sprintf("p99 %.0fms exceeded SLA of %dms", res.P99Latency, level.LatencySLA)
	case res.ErrorRate > level.ErrorSLAPct:
		res.Passed = false
		res.FailureReason = fmt.Sprintf("error rate %.1f%% exceeded SLA of %.1f%%", res.ErrorRate, level.ErrorSLAPct)
	case res.CPUPct >= 95.0:
		res.Passed = false
		res.FailureReason = fmt.Sprintf("CPU pinned at %.0f%% — server saturated", res.CPUPct)
	default:
		res.Passed = true
	}
	return res
}

func round1(x float64) float64 { return math.Round(x*10) / 10 }

File: services/constraint-engine/sim.py
from __future__ import annotations
import math
from model import Capabilities, ConstraintLevel, SimulationResult


def analyze_fix(fix: str) -> Capabilities:
    f = fix.lower()
    return Capabilities(
        cache="redis" in f or "cache" in f or "memcached" in f,
        persistent="redis" in f or "persist" in f or "external" in f
                    or "expire" in f or "ttl" in f or "database" in f or "db" in f,
        atomic="lua" in f or "incr" in f or "atomic" in f
                or "transaction" in f or "watch" in f or "compare-and" in f,
        pooling="pool" in f,
        replicas="replica" in f or "read replica" in f or "sentinel" in f,
        horizontal=any(kw in f for kw in ("lb", "load balancer", "load-balancer",
                    "scale out", "scale-out", "horizontal", "multiple servers", "autoscal")),
        sharded="shard" in f or "consistent hash" in f or "consistent-hash" in f or "partition" in f,
    )


def requirement_check(level: ConstraintLevel, fix: str) -> tuple[bool, str]:
    f = fix.lower()
    if level.requires:
        for req in level.requires:
            if req.lower() not in f:
                return False, req
    if level.requires_any:
        found = any(r.lower() in f for r in level.requires_any)
        if not found:
            return False, level.requires_any[0]
    return True, ""


def simulate(level: ConstraintLevel, caps: Capabilities) -> SimulationResult:
    rps = float(level.target_rps)
    single_node_cap = 2000.0
    effective_cap = single_node_cap
    if caps.horizontal:
        effective_cap *= 6
    if caps.sharded:
        effective_cap *= 1.5
    if caps.replicas:
        effective_cap *= 1.3

    util = rps / effective_cap if effective_cap > 0 else 1.0
    cpu = min(100, util * 100)

    base = 40.0
    if caps.cache:
        base = 8.0
    if caps.pooling:
        base -= 4
    base = max(base, 3)

    if util < 0.95:
        queue_factor = 1.0 / (1.0 - util)
    else:
        queue_factor = 20.0 + (util - 0.95) * 200.0

    p50 = base * min(queue_factor, 6)
    p99 = base * queue_factor * 2.5

    err_rate = 0.0
    if util > 1.0:
        err_rate += (util - 1.0) * 60.0
    if util > 0.85:
        err_rate += (util - 0.85) * 10.0
    err_rate = min(err_rate, 95.0)

    throughput = int(min(rps, effective_cap))

    return SimulationResult(
        p50_latency=round(p50, 1),
        p99_latency=round(p99, 1),
        error_rate=round(err_rate, 1),
        throughput=throughput,
        cpu_pct=round(cpu, 1),
    )


def evaluate(level: ConstraintLevel, fix: str) -> SimulationResult:
    caps = analyze_fix(fix)
    res = simulate(level, caps)

    ok, reason = requirement_check(level, fix)
    if not ok:
        res.passed = False
        res.failure_reason = f'missing required change: "{reason}" (need {level.change_type})'
        return res

    if res.p99_latency > level.latency_sla_ms:
        res.passed = False
        res.failure_reason = f"p99 {res.p99_latency:.0f}ms exceeded SLA of {level.latency_sla_ms}ms"
    elif res.error_rate > level.error_sla_pct:
        res.passed = False
        res.failure_reason = f"error rate {res.error_rate:.1f}% exceeded SLA of {level.error_sla_pct}%"
    elif res.cpu_pct >= 95.0:
        res.passed = False
        res.failure_reason = f"CPU pinned at {res.cpu_pct:.0f}% -- server saturated"
    else:
        res.passed = True
    return res
